internal/contentmanager/cache/cache.go

```diff
@@ -0,0 +1,240 @@
+	// Subsequent calls to Watch will result in no longer necessary
+	// watches being stopped and removed and new watches being
+	// created
+	Watch(context.Context, Watcher, ...client.Object) error
+}
+
+// CloserSyncingSource is a wrapper of the controller-runtime
+// source.SyncingSource that includes methods for:
+//   - Closing the source, stopping it's interaction with the Kubernetes API server and reaction to events
+type CloserSyncingSource interface {
+	source.SyncingSource
+	io.Closer
+}
+
+type sourcerer interface {
+	Source(schema.GroupVersionKind, client.Object, func(context.Context)) (CloserSyncingSource, error)
+}
+
+type cache struct {
+	sources     map[schema.GroupVersionKind]CloserSyncingSource
+	sourcerer   sourcerer
+	owner       client.Object
+	syncTimeout time.Duration
+	mu          sync.Mutex
+}
+
+func NewCache(sourcerer sourcerer, owner client.Object, syncTimeout time.Duration) Cache {
+	return &cache{
+		sources:     make(map[schema.GroupVersionKind]CloserSyncingSource),
+		sourcerer:   sourcerer,
+		owner:       owner,
+		syncTimeout: syncTimeout,
+	}
+}
+
+var _ Cache = (*cache)(nil)
+
+func (c *cache) Watch(ctx context.Context, watcher Watcher, objs ...client.Object) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	gvkSet, err := gvksForObjects(objs...)
+	if err != nil {
+		return fmt.Errorf("getting set of GVKs for managed objects: %w", err)
+	}
+
+	if err := c.removeStaleSources(gvkSet); err != nil {
+		return fmt.Errorf("removing stale sources: %w", err)
+	}
+	return c.startNewSources(ctx, gvkSet, watcher)
+}
+
+func (c *cache) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	errs := []error{}
+	for _, source := range c.sources {
+		if err := source.Close(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func (c *cache) startNewSources(ctx context.Context, gvks sets.Set[schema.GroupVersionKind], watcher Watcher) error {
+	cacheGvks := c.getCacheGVKs()
+	gvksToCreate := gvks.Difference(cacheGvks)
+
+	type startResult struct {
+		source CloserSyncingSource
+		gvk    schema.GroupVersionKind
+		err    error
+	}
+	startResults := make(chan startResult)
+	wg := sync.WaitGroup{}
+	for _, gvk := range gvksToCreate.UnsortedList() {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			source, err := c.startNewSource(ctx, gvk, watcher)
+			startResults <- startResult{
+				source: source,
+				gvk:    gvk,
+				err:    err,
+			}
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(startResults)
+	}()
+
+	sourcesErrors := []error{}
+	for result := range startResults {
+		if result.err != nil {
+			sourcesErrors = append(sourcesErrors, result.err)
+			continue
+		}
+
+		err := c.addSource(result.gvk, result.source)
+		if err != nil {
+			// If we made it here then there is a logic error in
+			// calculating the diffs between what is currently being
+			// watched by the cache
+			panic(err)
+		}
+	}
+
+	return errors.Join(sourcesErrors...)
```

Comment:
We'll get these `sourcesErrors` in random order because of the concurrent execution of goroutines that write to the `startResults` channel.

Should we sort these errors by their `Error() string` output (or maybe by their associated GVK?) in order to get a deterministic ordering (and resulting error string after being joined and returned)?

Reply:
IIRC GVKs are not sortable on their own (although maybe you could sort their string representation). I think the most straightforward approach here would probably be sorting on the error string output.

Reply:
Errors sorted by string in https://github.com/operator-framework/operator-controller/pull/1119/commits/f6f277035a8331898fd304504b30660905797406